dataprofiler/profilers/profile_builder.py

```diff
@@ -1511,23 +1511,80 @@ def save(self, filepath=None):
+class Profiler(object):
+
+    def __new__(cls, data, samples_per_update=None, min_true_samples=0,
+                options=None, profiler_type=None):
+        """
+        Factory class for instantiating Structured and Unstructured Profilers
+
+        :param data: Data to be profiled
+        :type data: Data class object
+        :param samples_per_update: Number of samples to use to generate profile
+        :type samples_per_update: int
+        :param min_true_samples: Min number of samples required for the profiler
+        :type min_true_samples: int
+        :param options: Options for the profiler.
+        :type options: ProfilerOptions Object
+        :param profiler_type: Type of Profiler ("structured"/"unstructured")
+        :type profiler_type: str
+        :return: BaseProfiler
+        """
+        # Attempt to construct according to user kwarg input
+        if profiler_type == "structured":
+            return StructuredProfiler(data, samples_per_update,
+                                      min_true_samples, options)
+        elif profiler_type == "unstructured":
+            return UnstructuredProfiler(data, samples_per_update,
+                                        min_true_samples, options)
+        elif profiler_type is not None:
+            raise ValueError("Must specify 'profiler_type' to be 'structured' "
+                             "or 'unstructured'.")
+
+        # If user doesn't specify, and data is Data object, use is_structured
+        if isinstance(data, data_readers.base_data.BaseData):
+            if data.is_structured:
+                return StructuredProfiler(data, samples_per_update,
+                                          min_true_samples, options)
+            else:
+                return UnstructuredProfiler(data, samples_per_update,
+                                            min_true_samples, options)
+
+        # If user doesn't specify, and data is Pandas object, need to infer
+        if isinstance(data, pd.DataFrame) or isinstance(data, pd.Series):
+            # If empty, default to structured
+            if not len(data):
+                warnings.warn("Empty data given to Profiler, will assume data "
+                              "is structured and create StructuredProfiler")
+                return StructuredProfiler(data, samples_per_update,
+                                          min_true_samples, options)
+
+            # If multi column df, data is structured
+            if isinstance(data, pd.DataFrame) and len(data.columns) > 1:
+                return StructuredProfiler(data, samples_per_update,
+                                          min_true_samples, options)
+
+            # If single column, look at string length and variance
+            data_ser = data if isinstance(data, pd.Series) else data[0]
+            if data_ser.dtype == "object":
+                sample = data_ser.sample(min(5, len(data_ser)))
+                sample_lens = sample.apply(len)
```

Comment:
before applying `len` we might want to do `astype(str)` as the objects may not be strings, they could be dicts

Reply:
Do the astype to `data` or `sample`?

Reply:
just the sample to save processing

Reply:
But what if data had mixed types, what is the proper behavior then?

Reply:
That's why we apply the astype(str) to unify it and just treat them as strings for now. If we want to get more complicated later we can.